File: Week Assignment 4 using C/1_insert_in_sorted_dll.c
// 1. Insert at beginning / Insert in sorted DLL (GFG style)
// We'll implement insertInSorted for a doubly linked list and also provide insertAtBeginning demo.
#include <stdio.h>
#include <stdlib.h>

struct Node { int data; struct Node* next; struct Node* prev; };

struct Node* insertAtBeginning(struct Node* head, int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = data; node->next = head; node->prev = NULL;
    if (head) head->prev = node;
    return node;
}

// Insert in sorted DLL (ascending)
struct Node* sortedInsert(struct Node* head, int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = data; node->next = node->prev = NULL;
    if (!head) return node;
    struct Node* cur = head;
    // insert at beginning
    if (data <= head->data) {
        node->next = head; head->prev = node; return node;
    }
    while (cur->next && cur->next->data < data) cur = cur->next;
    node->next = cur->next;
    if (cur->next) cur->next->prev = node;
    cur->next = node; node->prev = cur;
    return head;
}

void printListAsc(struct Node* head) {
    printf("[");
    struct Node* t = head; int first = 1;
    while (t) {
        if (!first) printf(", "); printf("%d", t->data); first = 0;
        t = t->next;
    }
    printf("]\n");
}

int main() {
    struct Node* head = NULL;
    head = insertAtBeginning(head, 30);
    head = insertAtBeginning(head, 20);
    head = insertAtBeginning(head, 10); // list: 10,20,30
    printListAsc(head);
    head = sortedInsert(head, 25); // expect [10, 20, 25, 30]
    printListAsc(head);
    head = sortedInsert(head, 5); // expect [5, 10, 20, 25, 30]
    printListAsc(head);
    return 0;
}

File: Week Assignement 4 using Python/1_insert_in_sorted_dll.py
# 1. Insert at beginning / insert in sorted DLL
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

def insert_at_beginning(head, data):
    n = Node(data)
    n.next = head
    if head: head.prev = n
    return n

def sorted_insert(head, data):
    n = Node(data)
    if not head: return n
    if data <= head.data:
        n.next = head; head.prev = n; return n
    cur = head
    while cur.next and cur.next.data < data: cur = cur.next
    n.next = cur.next
    if cur.next: cur.next.prev = n
    cur.next = n; n.prev = cur
    return head

def to_list(head):
    res = []
    cur = head
    while cur:
        res.append(cur.data); cur = cur.next
    return res

if __name__ == "__main__":
    head = None
    head = insert_at_beginning(head, 30)
    head = insert_at_beginning(head, 20)
    head = insert_at_beginning(head, 10)
    print(to_list(head))
    head = sorted_insert(head, 25); print(to_list(head))
    head = sorted_insert(head, 5); print(to_list(head))
